File: code/search/jump_search/jumpsearch.swift
import Foundation

func jumpSearch<T: Comparable>(_ a: [T], for key: T) -> Int? {
    let count = a.count
    var step = Int(sqrt(Double(count)))
    var previous = 0

    while a[min(step, count) - 1] < key {
        previous = step
        step += Int(sqrt(Double(count)))
        if previous >= count {
            return nil
        }
    }

    while a[previous] < key {
        previous += 1
        if previous == min(step, count) {
            return nil
        }
    }

    return a[previous] == key ? previous : nil
}


if let position = jumpSearch([7, 11, 13, 33, 66], for: 33) {
    print("Found 33 in [7, 11, 13, 33, 66] at position \(position)")
} else {
    print("Can't find 33 in [7, 11, 13, 33, 66]")
}

File: code/search/src/jump_search/jump_search.cs
using System; 
public class JumpSearch 
{ 
    public static int jumpSearch(int[] arr, int x) 
    { 
        int n = arr.Length; 
  
        // Finding block size to be jumped 
        int step = (int)Math.Floor(Math.Sqrt(n)); 
  
        // Finding the block where element is 
        // present (if it is present) 
        int prev = 0; 
        while (arr[Math.Min(step, n)-1] < x) 
        { 
            prev = step; 
            step += (int)Math.Floor(Math.Sqrt(n)); 
            if (prev >= n) 
                return -1; 
        } 
  
        // Doing a linear search for x in block 
        // beginning with prev. 
        while (arr[prev] < x) 
        { 
            prev++; 
  
            // If we reached next block or end of 
            // array, element is not present. 
            if (prev == Math.Min(step, n)) 
                return -1; 
        } 
  
        // If element is found 
        if (arr[prev] == x) 
            return prev; 
  
        return -1; 
    } 
  
    // Driver program to test function 
    public static void Main() 
    { 
        int[] arr = { 0, 1, 1, 2, 3, 5, 8, 13, 21, 
                    34, 55, 89, 144, 233, 377, 610}; 
        int x = 55; 
  
        // Find the index of 'x' using Jump Search 
        int index = jumpSearch(arr, x); 
  
        // Print the index where 'x' is located 
        Console.Write("Number " + x + 
                            " is at index " + index); 
    } 
} 
